

coexp.immune.cell.types = c("Bcells-External","Tcells-External","Thelpercells-External","Tcm-External","Tem-External",
                            "Th1cells-External","Th2cells-External","Tfh-External",
                            "CD8Tcells-External","Cytotoxiccells-External","NaturalKillers-External","iDC-External",
                            "Eosinophils-External","Macrophages-External","Mastcells-External","Neutrophils-External")

coexp.neuronal.cell.types = c("Neuron-In1","Neuron-In2","Neuron-In3","Neuron-In4",
                              "Neuron-In5","Neuron-In6","Neuron-In7","Neuron-In8",
                              "Neuron-Ex1","Neuron-Ex2","Neuron-Ex3","Neuron-Ex4",
                              "Neuron-Ex5","Neuron-Ex6","Neuron-Ex7","Neuron-Ex8")

coexp.cell.categories = c("Astrocyte_definite__Cahoy","brown_M15_Astrocyte__CTX",
                          "magenta_M8_Microglia(Type2)__HumanMeta","orange_M5_Microglia(Type2)__CTX",
                          "red_M11_Neuron__HumanMeta","blue_M16_Neuron__CTX",
                          "brown_pyramidalNeurons_Layer5/basolateralAmygdala__Sugino/Winden",
                          "pink_M10_Microglia(Type1)__HumanMeta","Neuron_definite__Cahoy",
                          "blue_M2_Oligodendrocytes__HumanMeta","turquoise_M9_Oligodendrocyte__CTX",
                          "Oligodendrocyte_definite__Cahoy","Oligodendrocyte__ABA","Neuron_probable__Cahoy")

#Pointers to the papers
coexp.long.friendly.cell.categories.ptp = 
  c("http://www.jneurosci.org/content/28/1/264.full.pdf+html", #Cahoy
    "http://www.nature.com/neuro/journal/v11/n11/full/nn.2207.html", #CTX
    "http://www.pnas.org/content/107/28/12698.abstract", #Human Meta
    "http://www.nature.com/neuro/journal/v11/n11/full/nn.2207.html", #CTX
    "http://www.pnas.org/content/107/28/12698.abstract", #Human Meta
    "http://www.nature.com/neuro/journal/v11/n11/full/nn.2207.html", #CTX
    "http://www.ncbi.nlm.nih.gov/pubmed/19638972", #Sugino/Winden
    "http://www.pnas.org/content/107/28/12698.abstract", #Human Meta
    "http://www.jneurosci.org/content/28/1/264.full.pdf+html", #Cahoy
    "http://www.pnas.org/content/107/28/12698.abstract", #Human Meta
    "http://www.nature.com/neuro/journal/v11/n11/full/nn.2207.html", #CTX
    "http://www.jneurosci.org/content/28/1/264.full.pdf+html", #Cahoy
    "http://www.nature.com/nature/journal/v445/n7124/abs/nature05453.html", #ABA
    "http://www.jneurosci.org/content/28/1/264.full.pdf+html" #Cahoy
  )


coexp.long.friendly.cell.categories = c("Astrocytes, highly probable (Cahoy, 2008)",
                                        "Astrocytes module in Cortex",
                                        "Microglia in Human brain Module (Geschwind, 2010)",
                                        "Microglia (Type2) module in Cortex",
                                        "Neuron in Human brain Module (Geschwind,2010)",
                                        "Neuron module in Cortex",
                                        "Neuron, pyramidal in network from Sugino/Winden",
                                        "Microglia (Type 1) (Geschwind, 2010)",
                                        "Neuron, definite (Cahoy, 2008)",
                                        "Oligodendrocytes in Human brain Module (Geschwind,2010)",
                                        "Oligodendrocytes in Cortex",
                                        "Oligodendrocytes, definite (Cahoy, 2008)",
                                        "Oligodendrocytes from conservative data set (Lein, 2007)",
                                        "Neuron, probably (Cahoy, 2008)")


coexp.friendly.cell.categories = c("Astrocytes-1",
                                   "Astrocytes-2",
                                   "Microglia",
                                   "Microglia (Type2)",
                                   "Neuron-1",
                                   "Neuron-2",
                                   "Pyramidal Neurons",
                                   "Microglia (Type1)",
                                   "Neuron-3",
                                   "Oligodendrocytes-1",
                                   "Oligodendrocytes-2",
                                   "Oligodendrocytes-3",
                                   "Oligodendrocytes-4",
                                   "Neuron (Prob)")

coexp.single.cell.types = c("Astrocytes",
                            "Astrocytes",
                            "Microglia",
                            "Microglia",
                            "Neuron",
                            "Neuron",
                            "Neuron",
                            "Microglia",
                            "Neuron",
                            "Oligodendrocytes",
                            "Oligodendrocytes",
                            "Oligodendrocytes",
                            "Oligodendrocytes",
                            "Neuron")

coexp.cell.types = c("astrocytes.1","astrocytes.2","microglia.1","microglia.2",
                     "dop. neuron.1","dop. neuron.2","pyramidal neurons","microglia.3",
                     "dop. neuron.3","oligodendrocytes.1","oligodendrocytes.2","oligodendrocytes.3",
                     "oligodendrocytes.4","dop. neuron 5")

coexp.cell.types.short = c("Astrocytes","Astrocytes","Microglia","Microglia",
                           "Neuron","Neuron","Pyramidal","Microglia",
                           "Neuron","Oligodendrocytes","Oligodendrocytes","Oligodendrocytes",
                           "Oligodendrocytes","Neuron")


coexp.ukbec.cell.types = c("Astrocytes-Cahoy","Neuron-Dopaminergic-Cahoy","Neurons-Cahoy","Oligodendrocytes-Cahoy")
coexp.ukbec.cell.files = c("astrocytes20.txt","dopaminergicneurons.txt","neurons20.txt","oligodendrocytes20.txt")

coexp.external.cell.types = c("Ependymal-Ext","Oligodendrocytes-Ext", "Microglia-Ext","CA1.Pyramidal-Ext", "Interneuron-Ext",     
                              "Endothelial", "S1.Pyramidal-Ext",  "Astrocytes-Ext", "Mural-Ext","Dopaminergic-Ext",
                              "SN-Dopaminergic","VTA-Dopaminergic",
                              "Neuron-In1","Neuron-In2","Neuron-In3","Neuron-In4","Neuron-In5","Neuron-In6","Neuron-In7","Neuron-In8",
                              "Neuron-Ex1","Neuron-Ex2","Neuron-Ex3","Neuron-Ex4","Neuron-Ex5","Neuron-Ex6","Neuron-Ex7","Neuron-Ex8",
                              "Bcells","Tcells","Thelpercells","Tcm","Tem","Th1cells","Th2cells","Tfh",
                              "CD8Tcells","Cytotoxiccells","NaturalKillers","iDC","Eosinophils",
                              "Macrophages","Mastcells","Neutrophils")

coexp.external.cell.types.short = c("Ependymal","Oligodendrocytes", "Microglia",
                                    "Pyramidal", "Interneuron",     
                                    "Endothelial", "Pyramidal",  
                                    "Astrocytes", "Mural","Neuron",
                                    "Neuron","Neuron",
                                    "Neuron","Neuron","Neuron","Neuron","Neuron","Neuron","Neuron","Neuron",
                                    "Neuron","Neuron","Neuron","Neuron","Neuron","Neuron","Neuron","Neuron",
                                    "Bcells","Tcells","Thelpercells","Tcm","Tem","Th1cells","Th2cells","Tfh",
                                    "CD8Tcells","Cytotoxiccells","NaturalKillers","iDC","Eosinophils",
                                    "Macrophages","Mastcells","Neutrophils")

coexp.reportExists = function(experiment,tissue,module){
  if(exists("coexp.report.go.all"))
    return(!is.null(coexp.report.go.all[[experiment]][[tissue]][[module]]))
  return(FALSE)	
}

coexp.reportGet = function(experiment,tissue,module){
  if(coexp.reportExists(experiment,tissue,module))
    return(coexp.report.go.all[[experiment]][[tissue]][[module]])
  return(NULL)
}

coexp.reportAdd = function(experiment,tissue,module,report){
  if(!exists("coexp.report.go.all")){
    coexp.report.go.all <<- NULL
    coexp.report.go.all[[experiment]] <<- NULL
    coexp.report.go.all[[experiment]][[tissue]] <<- NULL
    coexp.report.go.all[[experiment]][[tissue]][[module]] <<- report
  }else{
    if(is.null(report.go.all)){
      coexp.report.go.all[[experiment]] <<- NULL
      coexp.report.go.all[[experiment]][[tissue]] <<- NULL
      coexp.report.go.all[[experiment]][[tissue]][[module]] <<- report
    }else if(is.null(report.go.all[[experiment]])){
      coexp.report.go.all[[experiment]][[tissue]] <<- NULL
      coexp.report.go.all[[experiment]][[tissue]][[module]] <<- report
    }else #if(is.null(report.go.all[[experiment]][[tissue]])){
      coexp.report.go.all[[experiment]][[tissue]][[module]] <<- report
  }
}

coexp.reportOnFETGenes = function(tissue,genes,which.one,net){
  
  #Everything will be turned into Ensembl	
  if(is.null(net))
    net = getNetworkFromTissue(tissue=tissue,which.one=which.one)
  names(net$moduleColors) =  coexp.fromAny2Ensembl(names(net$moduleColors))
  en.genes = coexp.fromAny2Ensembl(genes)
  
  #Modules within genes
  mods = net$moduleColors[match(en.genes,names(net$moduleColors))]
  mods[is.na(mods)] = "void"
  final.report = NULL
  mult.mod.genes = NULL
  
  p.val.mods = rep(1,length(mods))
  table.mods = table(mods)
  table.mods = table.mods[names(table.mods) != "void"]
  total.specific = sum(mods != "void")
  total.net = length(net$moduleColors)
  for(module in names(table.mods)){
    n.module = sum(net$moduleColors == module)
    n.module.and.specific = table.mods[[module]]
    p.val = coexp.testGeneSet(n.module,n.module.and.specific,total.specific,total.net)
    p.val.mods[mods == module] = signif(p.val$p.value,4)
  }
  
  the.table = cbind(rep(tissue,length(genes)),rep(which.one,length(genes)),
                    genes,en.genes,mods,p.val.mods)
  colnames(the.table) = c("tissue","which.one","gene","ensgene","module","p.val.mods")
  rownames(the.table) = genes
  return(the.table)
}


coexp.reportOnGenes = function(tissue,genes,silent=F,
                               which.one="signedrnaseq",
                               mantel.its=1000,alt.probes=NULL,include.pd=T,
                               ens.correction=c("ENSG00000262446","ENSG00000177628",
                                                "ENSG00000261832", "ENSG00000188603",
                                                "ENSG00000262208", "ENSG00000114378",
                                                "ENSG00000228691","ENSG00000204386"),
                               gwases=NULL){
  
  #Everything will be turned into Ensembl	
  net = coexp.getNetworkFromTissue(tissue=tissue,which.one=which.one)
  
  names(net$moduleColors) =  coexp.fromAny2Ensembl(names(net$moduleColors))
  expr.data = coexp.getExprDataFromTissue(tissue=tissue,which.one=which.one)
  colnames(expr.data) = coexp.fromAny2Ensembl(colnames(expr.data))
  en.genes = coexp.fromAny2Ensembl(genes)
  
  if(!is.null(ens.correction)){
    for(i in seq(from=1,to=length(ens.correction),by=2)){
      mask = which(en.genes == ens.correction[i])
      if(sum(mask) > 0){
        en.genes[mask] = ens.correction[i+1]
        cat("Substituting", ens.correction[i],"by",ens.correction[i+1],"\n")	
      }
    }
  }
  
  final.report = NULL
  mult.mod.genes = NULL
  
  for(en.gene in en.genes){
    gene = en.gene
    cat("Working on gene",gene,genes[which(en.genes %in% gene)],"\n")
    mod = net$moduleColors[names(net$moduleColors) == en.gene]
    
    if(length(mod) >= 1){ #} | length(putm.mod) > 1){
      if(length(mod) > 1){
        mult.mod.genes = c(mult.mod.genes,en.gene)
        print(paste0("Gene ", en.gene," found in ",tissue," modules ",
                     mod))
        
        def.mod = NULL
        #Can we use probes to solve this?
        if(which.one == "micro19K"){
          if(!is.null(alt.probes)){
            #Lets try to find a probe for that
            probe.id.index = which(mic19k.probes %in% alt.probes[which(en.genes %in% gene)])
            def.mod = net$moduleColors[probe.id.index]
            cat("We will use",def.mod,"\n")
            mm = coexp.getMM(net=net,expr.data.file=expr.data,
                             genes=en.gene,alt.gene.index=probe.id.index)
          }
        }
        
        if(is.null(def.mod)){
          def.mod = mod[1]
          cat("We will use", gene,"found in module",def.mod,"in ",tissue,"\n")
          mod = paste0(mod,collapse=", ")
          mm = coexp.getMM(net=net,expr.data.file=expr.data,
                           genes=en.gene)
          
        }
        report = coexp.reportOnModule(tissue,def.mod,which.one=which.one,
                                      mantel.its=mantel.its,include.pd=include.pd)
        mod = def.mod
        
      }else{
        cat("Gene found in module",mod,"in ",tissue,"\n")
        report = coexp.reportOnModule(tissue,mod,which.one=which.one,
                                      mantel.its=mantel.its,include.pd=include.pd)
        mm = coexp.getMM(net=net,expr.data.file=expr.data,
                         genes=en.gene)
        
      }
      gene = coexp.fromAny2GeneName(gene)
      ensgene = en.gene
      
      report = c(gene = gene, 
                 ensgene = ensgene,
                 mm = signif(as.numeric(mm),4),
                 report)
      
    }else{
      cat("Gene not found in",tissue," gene set",which.one,"\n")
      mod = "void"
      mm = 0
      report = NULL
    }
    
    
    if(!silent){
      print(paste0("Gene ",gene, " is in ",tissue," in ",mod, " (MM ", 
                   mm,")"))
      print(paste0("Report for module ",mod))
      print(report)
    }
    
    #Get the evidence for cell specific genes based on WGCNA userLists
    wgcna.cell.cats = coexp.getFriendlyNameForCategories(coexp.getCategoriesForGene(gene))
    if(length(wgcna.cell.cats) == 0){
      wgcna.cell.cats = "void"
    }
    
    final.report[[gene]] = report
  }
  #And now we add the fisher exact test results
  
  #But before that we convert to dataframe
  the.table = NULL
  genes.with.report = names(final.report[unlist(lapply(final.report,function(x){
    return(!is.null(x))}))])
  for(gene in genes.with.report){
    report = final.report[[gene]]
    report$go.report = paste0(report$go.report,collapse=", ")
    report$pd.genes = paste0(report$pd.genes,collapse=", ")
    report.t = as.data.frame(report,stringsAsFactors=F)
    the.table = rbind(the.table,report.t)
  }
  
  #Now we add the VEGAS results if any
  if(!is.null(gwases)){
    modules = the.table$module
    vegastoadd = NULL
    for(module in modules){
      addbymodule = NULL
      for(gwas in gwases){
        newsignal = gsa.getSignal(which.one=which.one,tissue=tissue,gwas=gwas,module=module)
        if(!is.null(newsignal)){
          pvalue = newsignal$pvalue
          addbymodule = cbind(addbymodule,pvalue)
          colnames(addbymodule)[ncol(addbymodule)] = paste0(gwas,"vegaspvalue")
        }
        
      }
      if(!is.null(addbymodule)){
        vegastoadd = rbind(vegastoadd,addbymodule)
      }
      
    }
    if(!is.null(vegastoadd))
      the.table = cbind(the.table,vegastoadd)
  }
  
  total.specific = length(genes.with.report)
  total.net = length(net$moduleColors)
  mods = the.table$module
  
  #It could be that there are more than 1 modules per gene
  mult.mod.index = grep(",",mods)
  for(single.index in mult.mod.index){
    #Get the first module
    few.modules = mods[single.index]
    few.modules = str_split(few.modules,", ")
    mods[single.index] = few.modules[[1]][[1]]
  }
  
  if(include.pd & !is.null(mods)){
    p.val.mods = vector(mode="numeric",length=length(mods))
    names(p.val.mods) = mods
    table.mods = table(mods)
    table.mods = table.mods[names(table.mods) != "void"]
    for(module in names(table.mods)){
      n.module = sum(net$moduleColors == module)
      n.module.and.specific = table.mods[[module]]
      p.val = coexp.testGeneSet(n.module,n.module.and.specific,total.specific,total.net)
      p.val.mods[names(p.val.mods) == module] = signif(p.val$p.value,4)
    }
    the.table = cbind(the.table,p.val.mods)
    the.table = as.data.frame(the.table,stringsAsFactors=F)
    rownames(the.table) = names(final.report[genes.with.report])
    colnames(the.table[ncol(the.table)]) = "fisher"
    
  }
  cat("These genes are found in more than 1 module:",mult.mod.genes,"\n")
  return(list(report=the.table,mult.genes=mult.mod.genes))
}

coexp.reportOnGenesMultipleTissue = function(tissues,genes,silent=F,
                                             which.one="signedrnaseq",
                                             mantel.its=1000,
                                             alt.probes=NULL,
                                             out.file=NULL,
                                             include.pd=T,
                                             gwases=NULL,
                                             mod.level.correct=T){
  
  if(is.null(genes) & which.one == "micro19K")
    genes = names(getNetworkFromTissue(tissue="SNIG",which.one="micro19K")$moduleColors)
  
  out.table = NULL
  
  problematic.genes = NULL
  for(tissue in tissues){
    if(which.one == "gtexv6"){
      if(is.null(coexp.getNetworkFromTissue(which.one="gtexv6",tissue=tissue,only.file=T))){
        cat("Skipping",tissue,", no network available\n")
        next 
      }
      
    }
    tissue.report = coexp.reportOnGenes(tissue=tissue,
                                        genes=genes,silent=silent,
                                        which.one=which.one,
                                        mantel.its=mantel.its,alt.probes=alt.probes,include.pd=include.pd,gwases=gwases)
    if(!is.null(tissue.report$report)){
      problematic.genes[[tissue]] = tissue.report$mult.genes
      tissue.report = tissue.report$report
      tissue.report.ready = cbind(rep(tissue,nrow(tissue.report)),tissue.report)
      colnames(tissue.report.ready) = c("tissue",colnames(tissue.report))
      if(is.null(out.table)){
        out.table = tissue.report.ready
        
      }else{
        out.table = rbind(out.table,tissue.report.ready)
      }	
    }	
  }
  
  #We have to correct p.val.mods here!
  if(mod.level.correct){
    modules = 0
    for(tissue in tissues)
      modules = modules + length(coexp.getModulesFromTissue(tissue=tissue,
                                                            which.one=which.one)) 
    bonf = out.table$p.val.mods * modules
    bonf[bonf > 1] = 1
    out.table = cbind(out.table,bonf)
    
    colnames(out.table)[ncol(out.table)] = "bonfpval"
  }
  
  
  if(!is.null(out.file))
    write.csv(out.table,out.file)
  return(list(report=out.table,mult.genes=problematic.genes))
}

coexp.functionalReportOnModule = function(tissue="SNIG",module,which.one="rnaseq"){
  
  if(which.one  == "micro19K"){
    
    gprof.file = paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/micro19K/"),
                        tissue,"_cor_pca_GO.csv")
    
  }else if(which.one == "exonic"){
    if(tissue == "SNIG")
      gprof.file = paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/netSNIG.7.12.it.30.rds_gprofiler.csv")
    else
      gprof.file = paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/netPUTM.8.13.it.30.rds_gprofiler.csv")
    
  }else if(which.one == "gtexv6"){
    gprof.file = paste0(coexp.getNetworkFromTissue(tissue=tissue,which.one="gtexv6",only.file=T),"_gprof.csv")
    
  }else if(which.one == "rosmap"){
    if(tissue == "cogdxnotad")
      gprof.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap/netnotad.8.it.50.rds_gprof.csv")
    else if(tissue == "cogdxprobad")
      gprof.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap/netprobad.11.it.50.rds_gprof.csv")
    else if(tissue == "cogdxad")
      gprof.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap//netad.8.it.50.rds_gprof.csv")
    else if(tissue == "all")
      gprof.file = paste0(gdp.coexp(),"//supplementary/rdsnets/rosmap/netROSMAPSingle.6.it.50.rds_gprof.csv")
  }
  
  
  go <- read.csv(gprof.file,stringsAsFactors=FALSE)
  go$X = NULL
  return(go[go$query.number == module,])
}

coexp.reportOnModule = function(tissue="SNIG",module,which.one="rnaseq",how.many=5,
                                include.pd=T,mantel.its=1000,simple.cell.types=F,
                                cell.type.threshold=0.05,ctcollapse=T){
  
  tmp.report = coexp.reportGet(which.one,tissue,module)
  if(!is.null(tmp.report)){
    #cat("Report for",module,which.one,tissue,"exists already\n")
    return(tmp.report)
  }
  #cat("Report for",module, which.one,tissue,"doesn't exist already\n")
  
  if(which.one  == "micro19K"){
    
    gprof.file = paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/micro19K/"),
                        tissue,"_cor_pca_GO.csv")
    
  }else if(which.one == "exonic"){
    if(tissue == "SNIG")
      gprof.file = paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/netSNIG.7.12.it.30.rds_gprofiler.csv")
    else
      gprof.file = paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/netPUTM.8.13.it.30.rds_gprofiler.csv")
    
    mt.between = readRDS(paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/mantel."),tissue,
                                ".gtex.binary.between.",mantel.its,".data.rds"))
    
    mt.within = readRDS(paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/mantel."),tissue,
                               ".gtex.coexpression.within.",mantel.its,".data.rds"))
    mt.mic.between = readRDS(paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/mantel."),tissue,
                                    ".microarray.binary.between.",mantel.its,".data.rds"))
    mt.mic.within = readRDS(paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/mantel."),tissue,
                                   ".microarray.coexpression.within.",mantel.its,".data.rds"))
  }else if(which.one == "gtexv6"){
    gprof.file = paste0(coexp.getNetworkFromTissue(tissue=tissue,which.one="gtexv6",only.file=T),"_gprof.csv")
    
  }else if(which.one == "rosmap"){
    if(tissue == "cogdxnotad")
      gprof.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap/netnotad.8.it.50.rds_gprof.csv")
    else if(tissue == "cogdxprobad")
      gprof.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap/netprobad.11.it.50.rds_gprof.csv")
    else if(tissue == "cogdxad")
      gprof.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap//netad.8.it.50.rds_gprof.csv")
    else if(tissue == "all")
      gprof.file = paste0(gdp.coexp(),"/supplementary/rdsnets/rosmap/netROSMAPSingle.6.it.50.rds_gprof.csv")
  }
  
  
  pd.genes = "void"
  if(include.pd){
    
    pd.genes = read.table(paste0(gdp.coexp(),"/data/pd_genes.txt"),
                          stringsAsFactors=F,header=F)
    pd.genes = pd.genes$V1
    genes.in.module = coexp.getGenesFromModule(tissue=tissue,which.one=which.one,module=module)
    if(substr(genes.in.module[1],1,4) == "ENSG")
      genes.in.module = coexp.fromEnsembl2GeneName(genes.in.module)
    mask = pd.genes %in% genes.in.module
    if(sum(mask) > 0)
      pd.genes = pd.genes[mask]
    else
      pd.genes = "void"
  }
  
  go <- read.csv(gprof.file,stringsAsFactors=FALSE)
  
  p.values <- go$p.value[go$query.number == module & (go$domain %in% "BP")]
  terms <- go$term.name[go$query.number == module & (go$domain %in% "BP")]
  term.ids = go$term.id[go$query.number == module & (go$domain %in% "BP")]
  
  if("IC" %in% colnames(go)){
    ics = go$IC[go$query.number == module & (go$domain %in% "BP")]
    p.values = p.values[!is.na(ics)]
    terms = terms[!is.na(ics)]
    ics = na.omit(ics)
  }
  
  else
    ics = NULL
  go.report = ""
  if(length(p.values) > 0){
    go.count = length(p.values)
    if(how.many > go.count)
      how.many = go.count 
    the.order <- order(-log10(p.values),decreasing=TRUE)
    if(is.null(ics))
      go.report = paste0(terms[the.order][1:how.many]," ",term.ids[the.order][1:how.many], 
                         " (p-value ",signif(p.values[the.order][1:how.many],4),")")
    else{
      p.values = p.values[the.order]
      terms = terms[the.order]
      term.ids = term.ids[the.order]
      ics = ics[the.order]
      
      p.values.f = p.values[ics > 2.5]
      terms.f = terms[ics > 2.5]
      term.ids.f = term.ids[ics > 2.5]
      
      if(length(p.values.f) > 0){
        if(length(p.values.f) < how.many)
          how.many = length(p.values.f)
        go.report = paste0(terms.f[1:how.many]," ",term.ids.f[1:how.many], 
                           " (p-value ",signif(p.values.f[1:how.many],4),")")
      }else
        go.report = paste0(terms[1:how.many], " ",term.ids.f[1:how.many], 
                           " (p-value ",signif(p.values[1:how.many],4),")")	
    }
    go.report.size = go.count
    go.report.signif = sum(-log10(p.values))
  }else{
    go.report = "void"
    go.report.size = go.report.signif = 0
  }
  
  if(which.one  == "exonic"){
    
    #We include cell types here
    cell.types = coexp.cellTypeByModule(tissue=tissue,do.plot=F,which.one=which.one,
                                        return.processed=F)
    cell.type.report = "void"
    if(any(colnames(cell.types) %in% module)){
      cell.types = cell.types[order(cell.types[,module],decreasing=F),]
      cell.type.row = which(cell.types[,module] <= cell.type.threshold)
      if(length(cell.type.row) > 0){
        cell.type.report = NULL
        if(!simple.cell.types){
          
          for(i in 1:length(cell.type.row)){
            if(ctcollapse)
              cell.type.report = paste0(cell.type.report," ",
                                        coexp.getFriendlyNameForColumnCategories(names(cell.type.row)[i]),
                                        " (p-value ",signif(cell.types[cell.type.row[i],module],4),"). ")
            else
              cell.type.report = c(cell.type.report,
                                   paste0(coexp.getFriendlyNameForColumnCategories(names(cell.type.row)[i]),
                                          " (p-value ",signif(cell.types[cell.type.row[i],module],4),")"))
            
          }
        }else{
          cell.type.report = unique(coexp.getFriendlyNameForColumnCategories(names(cell.type.row),F))
        }
      }
    }
    
    replication = coexp.reportOnModuleReplication(tissue=tissue,module=module)
    
    
    to.return = list(module=module,
                     size=length(coexp.getGenesFromModule(tissue,which.one,module)),
                     go.report=go.report,
                     go.report.size=go.report.size,
                     go.report.signif=go.report.signif,
                     Z.summary=replication["Z.summary"],
                     Z.summary.GTEx=replication["Z.summary.GTEx"],
                     mantel.within=replication["mantel.GNAT.within"],
                     mantel.between=replication["mantel.GNAT.between"],
                     mantel.micro.within=replication["mantel.micro.within"],
                     mantel.micro.between=replication["mantel.micro.between"],
                     pd.genes = pd.genes,
                     cell.type.pred=cell.type.report)
    null.values.mask = unlist(lapply(to.return,function(x){return(is.null(x))}))
    to.return[null.values.mask] = "Not available"
    coexp.reportAdd(which.one,tissue,module,to.return)
    return(to.return)
    
    
    to.return = list(module=module,
                     size=length(coexp.getGenesFromModule(tissue,which.one,module)),
                     go.report=go.report,
                     go.report.size=go.report.size,
                     go.report.signif=go.report.signif,
                     gtex.within=mt.within.p.vals,
                     gtex.between=mt.between.p.vals,
                     micro.within=mt.mic.within.p.vals,
                     micro.between=mt.mic.between.p.vals,
                     pd.genes = pd.genes,
                     preservation = coexp.getZSummary(tissue=tissue,module=module,which.one=which.one),
                     cell.type.pred=cell.type.report)
    null.values.mask = unlist(lapply(to.return,function(x){return(is.null(x))}))
    to.return[null.values.mask] = "Not available"
    coexp.reportAdd(which.one,tissue,module,to.return)
    return(to.return)
  }
  #We include cell types here
  cell.types = coexp.cellTypeByModule(tissue=tissue,do.plot=F,which.one=which.one,
                                      return.processed=F)
  cell.type.report = "void"
  if(any(colnames(cell.types) %in% module)){
    cell.types = cell.types[order(cell.types[,module],decreasing=F),]
    cell.type.row = which(cell.types[,module] <= cell.type.threshold)
    if(length(cell.type.row) > 0){
      cell.type.report = NULL
      if(!simple.cell.types){
        
        for(i in 1:length(cell.type.row)){
          if(ctcollapse)
            cell.type.report = paste0(cell.type.report," ",
                                      coexp.getFriendlyNameForColumnCategories(names(cell.type.row)[i]),
                                      " (p-value ",signif(cell.types[cell.type.row[i],module],4),"). ")
          else
            cell.type.report = c(cell.type.report,
                                 paste0(coexp.getFriendlyNameForColumnCategories(names(cell.type.row)[i]),
                                        " (p-value ",signif(cell.types[cell.type.row[i],module],4),")"))
          
        }
      }else{
        cell.type.report = unique(coexp.getFriendlyNameForColumnCategories(names(cell.type.row),F))
      }
    }
  }
  if(which.one == "micro19K"){
    preservation = coexp.getZsummary10MicTissues1Module(tissue=tissue,
                                                        module=module)
    
    to.return = list(module=module,
                     size=length(coexp.getGenesFromModule(tissue=tissue,which.one=which.one,
                                                          module=module)),
                     go.report=go.report,
                     pd.genes=pd.genes,
                     min.prev=preservation$min,
                     max.prev=preservation$max,
                     mean.prev=preservation$mean,
                     cell.type.pred=cell.type.report)
    coexp.reportAdd(which.one,tissue,module,to.return)
    return(to.return)
  }else{
    preservation = coexp.getZSummary(tissue=tissue,module=module,which.one=which.one)
    if(is.na(preservation))
      preservation = "void"
    to.return = list(module=module,
                     size=length(coexp.getGenesFromModule(tissue=tissue,which.one=which.one,
                                                          module=module)),
                     go.report=go.report,
                     pd.genes=pd.genes,
                     preservation=preservation,
                     cell.type.pred=cell.type.report)
    coexp.reportAdd(which.one,tissue,module,to.return)
    return(to.return)
    
  }
}

coexp.reportOnModuleReplication = function(tissue="SNIG",module,
                                           which.one="exonic",mantel.its=1000){
  
  #Note that only the combination binary-between and co-expression within make sense
  #The other two are not good
  
  #Mantel results
  stopifnot(which.one == "exonic")
  mt.between = readRDS(paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/mantel."),tissue,
                              ".gtex.binary.between.",mantel.its,".data.rds"))
  mt.within = readRDS(paste0(paste0(gdp.coexp(),"//supplementary/rdsnets/exonic/mantel."),tissue,
                             ".gtex.coexpression.within.",mantel.its,".data.rds"))
  mt.mic.between = readRDS(paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/mantel."),tissue,
                                  ".microarray.binary.between.",mantel.its,".data.rds"))
  mt.mic.within = readRDS(paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/mantel."),tissue,
                                 ".microarray.coexpression.within.",mantel.its,".data.rds"))
  
  if(module %in% names(mt.within))
    mt.within.p.vals = mt.within[[module]]$p
  else
    mt.within.p.vals = "void"
  if(module %in% names(mt.between))
    mt.between.p.vals = mt.between[[module]]$p
  else 
    mt.between.p.vals = "void"
  if(module %in% names(mt.mic.within))
    mt.mic.within.p.vals = mt.mic.within[[module]]$p
  else
    mt.mic.within.p.vals = "void"
  if(module %in% names(mt.mic.between))
    mt.mic.between.p.vals = mt.mic.between[[module]]$p
  else
    mt.mic.between.p.vals = "void"
  preservation = coexp.getZSummary(tissue=tissue,module=module,which.one=which.one)
  pres.gtex = coexp.getZSummary(tissue=tissue,module=module,which.one="exonic.vs.gtex")
  pres.micro = coexp.getZSummary(tissue=tissue,module=module,which.one="exonic.vs.micro")
  
  if(is.na(preservation))
    preservation = "void"
  to.return = c(preservation,pres.gtex,pres.micro,
                mt.within.p.vals,mt.between.p.vals,mt.mic.within.p.vals,mt.mic.between.p.vals)
  names(to.return) = c("Z.summary","Z.summary.GTEx","Z.summary.micro","mantel.GNAT.within","mantel.GNAT.between",
                       "mantel.micro.within","mantel.micro.between")
  return(to.return)	
}

coexp.getZSummary = function(tissue="SNIG",
                             modules=c("red"),
                             which.one="signedrnaseq"){
  
  if(is.null(modules))
    modules = unique(coexp.getNetworkFromTissue(tissue=tissue,which.one=which.one)$moduleColors)
  
  source = 1
  target = 2
  if(which.one == "micro19K")
  {
    source = 2
    target = 1
    pr.file = paste0(gdp.coexp(),
                     "/supplementary/rdsnets/micro19K/PUTM.mic.net.19K.rds_vs_SNIG.mic.net.19K.rds_preserv.rds")
  }else if(which.one == "exonic"){
    pr.file = paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/SNIG_vs_PUTM_preserv.rds")
  }else if(which.one == "exonic.vs.gtex"){
    pr.file = paste0(gdp.coexp(),"/supplementary/rdsnets/gtexv6/",
                     tissue,".UKBEC_vs_",tissue,".GTEx_preserv.rds")
    stats = readRDS(pr.file)
    loc.modules = rownames(stats$preservation$Z[[1]][[2]])
    return(stats$preservation$Z[[1]][[2]][match(modules,loc.modules),"Zsummary.pres"])
  }else if(which.one == "exonic.vs.micro"){
    pr.file = paste0(gdp.coexp(),"/supplementary/rdsnets/exonic/",
                     tissue,"-RNAseq_vs_",tissue,"-microarray_preserv.rds")
    stats = readRDS(pr.file)
    loc.modules = rownames(stats$preservation$Z[[1]][[2]])
    return(stats$preservation$Z[[1]][[2]][match(modules,loc.modules),"Zsummary.pres"])
  }else if(which.one == "gtexv6"){
    return(NA)
  }else if(which.one == "rosmap"){
    return(NA)
  }else{
    stop(paste0("In getZSummary which.one=",which.one," not known"))
  }
  
  stats = readRDS(pr.file)
  
  if(tissue == "SNIG"){
    loc.modules = rownames(stats$preservation$Z[[source]][[target]])
    return(stats$preservation$Z[[source]][[target]][match(modules,loc.modules),"Zsummary.pres"])
  }else if (tissue == "PUTM"){
    loc.modules = rownames(stats$preservation$Z[[target]][[source]])
    return(stats$preservation$Z[[target]][[source]][match(modules,loc.modules),"Zsummary.pres"])
  }else return(NA)
}


coexp.getZsummary10MicTissues1Module = function(tissue,module){
  tissues = coexp.getMicTissues()
  i = which(tissues == tissue)
  if(i == 1)
    lefttissues = NULL
  else
    lefttissues = tissues[1:(i - 1)]
  if(i == 10)
    righttissues = NULL
  else
    righttissues = tissues[(i+1):10]
  prvals = NULL
  
  tnet = coexp.getNetworkFromTissue(tissue=tissue,which.one="micro19K")
  outtissues = NULL
  for(othertissue in lefttissues){
    othernet = coexp.getNetworkFromTissue(tissue=othertissue,which.one="micro19K")
    file.in = paste0(gdp.coexp(),"/supplementary/rdsnets/micro19K/",
                     othertissue,".mic.net.19K.rds_vs_",tissue,
                     ".mic.net.19K.rds_preserv.rds","Zsummary.csv")
    if(!file.exists(file.in))
      coexp.getPreservationStatistics(networks=list(othernet,tnet),
                                      tissues=c(othertissue,tissue),
                                      paste0(gdp.coexp(),"/supplementary/rdsnets/micro19K/",othertissue,".mic.net.19K.rds_vs_",tissue,
                                             ".mic.net.19K.rds_preserv.rds"))
    
    pr = read.csv(file=file.in,stringsAsFactors=F)
    if(sum(pr$tissue == tissue & pr$color == module) == 0)
      stop(paste0("When checking preservation of ",module," from tissue ",tissue,
                  " at tissue ",othertissue," we cant find the module"))
    prvals = c(prvals,pr[pr$tissue == tissue & pr$color == module,othertissue])	
    outtissues = c(outtissues,othertissue)
  }
  
  for(othertissue in righttissues){
    othernet = coexp.getNetworkFromTissue(tissue=othertissue,which.one="micro19K")
    file.in = paste0(gdp.coexp(),"/supplementary/rdsnets/micro19K/",
                     tissue,".mic.net.19K.rds_vs_",othertissue,
                     ".mic.net.19K.rds_preserv.rds","Zsummary.csv")
    if(!file.exists(file.in))
      coexp.getPreservationStatistics(networks=list(tnet,othernet),
                                      tissues=c(tissue,othertissue),
                                      paste0(gdp.coexp(),"/supplementary/rdsnets/micro19K/",
                                             tissue,".mic.net.19K.rds_vs_",othertissue,
                                             ".mic.net.19K.rds_preserv.rds"))
    pr = read.csv(file=file.in,stringsAsFactors=F)
    if(sum(pr$tissue == tissue & pr$color == module) == 0)
      stop(paste0("When checking preservation of ",module," from tissue ",tissue,
                  " at tissue ",othertissue," we cant find the module"))
    prvals = c(prvals,pr[pr$tissue == tissue & pr$color == module,othertissue])
    outtissues = c(outtissues,othertissue)
  }
  names(prvals) = c(lefttissues,righttissues)
  return(list(min=min(prvals),max=max(prvals),mean=mean(prvals),all=prvals,tissues=outtissues))
}

coexp.getPreservationStatistics = function(networks,tissues,prev.data.file){
  
  net.objects <- list()
  
  if(typeof(networks[[1]]) == "character"){
    for(index in 1:length(networks)){
      print(paste0("Loading network ",networks[index]," for tissue ",tissues[index]))
      net.objects[[tissues[index]]] <- readRDS(networks[[index]])	
    }
  }else{
    for(index in 1:length(networks)){
      print(paste0("Loading network for tissue ",tissues[index]))
      net.objects[[tissues[index]]] <- networks[[index]]
    }
    
  }
  
  
  
  Zsummary <- NULL
  MedianRank <- NULL
  
  for(tissue1 in tissues){
    Z.tmp.list <- list(NULL)
    MR.tmp.list <- list(NULL)
    for(tissue2 in tissues){
      ## loop through columns
      cat(tissue1, "\t", tissue2, "\n")
      if(tissue2 == tissue1) next() ## skip self-self comparison
      if( tissue1 < tissue2 ){
        fn	<- paste(tissue1, "vs", tissue2,sep="")
        ref <- 1
        test <- 2
      } else {
        ## swap around
        fn <- paste(tissue2, "vs", tissue1,sep="")
        ref <- 2
        test <- 1
      }
      mp = readRDS(prev.data.file)
      ### other statistics can be exstracted here
      #these statistics for instance can be exstracted:
      #Zsummary.pres","Zconnectivity.pres","Zdensity.pres","Z.cor.kIM"
      
      Z.tmp <- mp$preservation$Z[[ref]][[test]][,"Zsummary.pres",drop=F]
      rownames(Z.tmp) <- paste(tissue1, "_", rownames(Z.tmp),sep="")
      colnames(Z.tmp) <- tissue2
      Z.tmp.list[[ tissue2 ]] <- Z.tmp
      
      MR.tmp <- mp$preservation$observed [[ ref ]][[ test ]][ , "medianRank.pres", drop=F]
      rownames(MR.tmp) <- paste(tissue1, "_", rownames(MR.tmp),sep="")
      colnames(MR.tmp) <- tissue2
      MR.tmp.list[[ tissue2 ]] <- MR.tmp
      rm(mp, Z.tmp, MR.tmp, tissue2, fn, ref, test)
    }
    
    Z.tmp.list <- Z.tmp.list[ which(names(Z.tmp.list) != "") ]
    Z.tmp.list <- cbind( do.call("cbind", Z.tmp.list), tissue1=NA )
    colnames(Z.tmp.list)[ colnames(Z.tmp.list)=="tissue1" ] <- tissue1
    
    Zsummary <- rbind(Zsummary, Z.tmp.list[ , tissues])
    MR.tmp.list <- MR.tmp.list[ which(names(MR.tmp.list) != "") ]
    MR.tmp.list <- cbind( do.call("cbind", MR.tmp.list), tissue1=NA )
    colnames(MR.tmp.list)[ colnames(MR.tmp.list)=="tissue1" ] <- tissue1
    MedianRank <- rbind(MedianRank, MR.tmp.list[ , tissues])
  }
  
  # 230 i.e. ten extra
  # gold module consists of 1000 randomly selected genes
  print(dim(Zsummary))
  w <- grep("gold", rownames(Zsummary))
  round( Zsummary[w, ], 2 )
  
  MedianRank[w, ]
  Zsummary <- Zsummary[ -w, ]
  MedianRank <- MedianRank[ -w, ]
  
  # get the real module sizes
  moduleColors <- sapply(net.objects, function(obj) obj$moduleColors)
  mat = NULL
  mat[[tissues[1]]] = table(moduleColors[,1])
  mat[[tissues[2]]] = table(moduleColors[,2])
  
  #mat <- apply( moduleColors[, 1:length(tissues)], 2, function(x) table(x) )
  mat <- data.frame(as.matrix(unlist(mat)))
  rownames(mat) <- gsub("[.]","_",rownames(mat))
  kk <- match(rownames(mat),rownames(Zsummary))
  Zsummary <- Zsummary[kk,]
  kk <- match(rownames(mat),rownames(MedianRank))
  MedianRank <- MedianRank[kk,]
  
  Zsummary <- cbind( mat[,1],Zsummary)
  colors.col <- NULL
  tissue.col <- NULL
  for(index in 1:length(rownames(mat))){
    values <- str_split(rownames(mat)[index],"_")
    colors.col[index] <- values[[1]][2]
    tissue.col[index] <- values[[1]][1]
  }
  Zsummary <- cbind(tissue.col,colors.col,Zsummary)
  colnames( Zsummary ) <- c("tissue","color","size",tissues)
  
  MedianRank <- cbind( mat[,1],MedianRank )
  colnames( MedianRank ) <- c("size",tissues)
  write.csv( Zsummary, file=paste0(prev.data.file,"Zsummary.csv"), quote=F )
  write.csv( MedianRank, file=paste0(prev.data.file,"MedianRank.csv"), quote=F )
  
}


coexp.UserGOenrichment <- function(net.file,net.name=NULL,which.one="exonic",do.plot=F){
  #Variable initialization
  if(typeof(net.file) == "character")
    net <- readRDS(net.file)
  else
    net = net.file
  if(is.null(net.name))
    net.name <- str_extract(net.file,"network[0-9]+.+rds$")
  out.file <- paste0(net.name,".USER_terms.csv")
  
  cat("Passing",length(net$moduleColors)," gene symbols to user enrichment\n")
  
  cat("Writting everything to",out.file,"\n")
  
  if(which.one != "braineac")
    genes = coexp.fromAny2GeneName(names(net$moduleColors))
  else
    genes = coexp.newmic.fromtID2GeneName(names(net$moduleColors))
  
  enrichments = userListEnrichment(genes,
                                   net$moduleColors, 
                                   fnIn=NULL, 
                                   useBrainLists = TRUE,
                                   useBrainRegionMarkers=T,
                                   useImmunePathwayLists=T,
                                   useBloodAtlases=T,
                                   useStemCellLists=T,
                                   usePalazzoloWang=T,
                                   nameOut=out.file, outputCorrectedPvalues=TRUE)
  
  
  
  if(do.plot){
    pdf(paste0(net.file,".go_user_zscores.pdf"),width=8,height=6)
    coexp.plotGOZScoresUserList(enrichments,net,title=paste0("User enrichment. Sum z-s (p-val < 10e-2), ",
                                                             net.name))
    dev.off()
  }
  write.csv(enrichments$sigOverlaps,out.file)
  #saveGOTermDefinition(enrichments,out.file)
  
}

#This function generates a csv file with the Gene Ontology enrichment
#signals. It is based on the gProfileR R package (see documentation for the
#package). Main parameters are
#net.file			: full path name for the net file
#filter				: ontologies to test for enrichemtn (see gProfileR docs)
#exclude.iea		: do not use Inferred electronic annotations (see gProfileR docs)
#correction.method	: method for multiple testing correction (see gProfileR docs)
#out.file			: full name for the csv file where to store results
#incluster			: ignore that one
coexp.getGProfilerOnNet <- function(net.file=snig.exonic.net.file,
                                    filter=c("GO","KEGG","REAC"),
                                    ensembl=TRUE,
                                    exclude.iea=T,
                                    correction.method="gSCS",
                                    out.file=paste0(net.file,"_gprofiler.csv")){
  
  if(typeof(net.file) == "character")
    net <- readRDS(net.file)
  else
    net = net.file
  
  modules = unique(net$moduleColors)
  background <- names(net$moduleColors)
  if(ensembl)
    background <- coexp.fromEnsembl2GeneName(background)
  
  all.genes <- NULL
  for(module in modules){
    
    genes <- names(net$moduleColors)[net$moduleColors == module]
    if(ensembl)
      genes <- coexp.fromEnsembl2GeneName(genes)
    all.genes[[module]] <- genes
  }
  
  go <- gprofiler(all.genes,correction_method=correction.method,
                  #custom_bg=background,
                  src_filter=filter,
                  exclude_iea=exclude.iea)
  #png_fn = paste0(out.file,".png"),
  #no_isects=T)
  go.out = cbind(go,rep(NA,nrow(go)))
  colnames(go.out) = c(colnames(go),"IC")
  
  #	#Now we will add our own column with the information content of each term
  #	#So it is possible to evaluate them
  coexp.loadICsGOSim("BP")
  mask = go$domain == "BP"
  bp.terms.go = go$term.id[mask]
  bp.ic.go = IC[bp.terms.go]
  go.out$IC[mask] = bp.ic.go
  
  coexp.loadICsGOSim("MF")
  mask = go$domain == "MF"
  mf.terms.go = go$term.id[mask]
  mf.ic.go = IC[mf.terms.go]
  go.out$IC[mask] = mf.ic.go
  
  coexp.loadICsGOSim("CC")
  mask = go$domain == "CC"
  cc.terms.go = go$term.id[mask]
  cc.ic.go = IC[cc.terms.go]
  go.out$IC[mask] = cc.ic.go
  #	
  #	data("ICsMFhumanall")
  #	data("ICsCChumanall")
  if(!is.null(out.file)){
    write.csv(go.out,out.file)
    cat("Obtained",nrow(go),"terms saved at",out.file,"\n")
  }else{
    cat("Obtained",nrow(go),"terms\n")
  }
  return(go.out)
}

coexp.loadICsGOSim = function(onto){
  stopifnot(onto %in% c("BP","MF","CC"))
  path = paste0(gdp.coexp(),"/data/")
  if(onto == "BP")
    load(paste0(path,"ICsBPhumanall.rda"),
         .GlobalEnv)
  else if(onto == "MF")
    load(paste0(path,"ICsMFhumanall.rda"),
         .GlobalEnv)
  else
    load(paste0(path,"ICsCChumanall.rda"),
         .GlobalEnv)
}

coexp.getICReport = function(gprof){
  ontologies = c("BP","CC","MF")
  ics = NULL
  for(onto in ontologies){
    loadICsGOSim(onto)
    terms = gprof$term.id[gprof$domain == onto]
    ics[[onto]] = IC[terms]
    names(ics[[onto]]) = terms
  }
  return(ics)
}

coexp.plotGOZScoresUserList <- function(goresult,net,title="Sum z-scores "){
  #Transforming p values in log10 scale
  logs <- -log10(goresult$sigOverlaps$CorrectedPvalues)
  modules <- unique(goresult$sigOverlaps$InputCategories)
  
  mod.colors <- as.character(modules)
  
  print(modules)
  z.scores <- NULL
  n.terms <- NULL
  mod.labels <- NULL
  c.modules <- 0
  
  
  for(module in modules){
    indexes <- which(goresult$sigOverlaps$InputCategories == module)
    print(indexes)
    good.p.values <- logs[indexes]
    c.modules <- c.modules + 1
    z.scores[c.modules] <- sum(good.p.values)
    n.terms[c.modules] <- length(good.p.values)
    mod.labels[c.modules] <- paste0(n.terms[c.modules],"/",
                                    table(net$moduleColors == module)[2])
  }	
  
  data.to.plot <- as.data.frame(cbind(1:c.modules,z.scores,mod.colors,mod.labels,n.terms),
                                stringsAsFactors=FALSE)	
  data.to.plot <- data.to.plot[order(-z.scores,-n.terms),]
  
  x <- barplot(height=as.numeric(data.to.plot$z.scores),names.arg=data.to.plot$mod.labels,
               width=as.numeric(data.to.plot$n.terms),col=data.to.plot$mod.colors,
               main=title, xaxt="n",
               ylab="-log10(pval)")
  text(cex=1, x=x-2, y=-1.25, pos=1, offset=2, data.to.plot$mod.labels, xpd=TRUE, srt=45)
  legend("topright", title="Module colors", 
         legend=data.to.plot$mod.colors,
         fill=data.to.plot$mod.colors,cex=0.5) 	
}

coexp.getFriendlyNameForCategories = function(cats){
  friendly.cats = cats
  existing.cats = cats[cats %in% coexp.cell.categories]
  friendly.cats[cats %in% coexp.cell.categories] = 
    coexp.long.friendly.cell.categories[match(existing.cats,coexp.cell.categories)]
  return(friendly.cats)
}

coexp.getCategoriesForGene = function(gene){
  #Load the gene lists
  if (!(exists("BrainLists"))) 
    BrainLists = NULL
  data("BrainLists", envir = sys.frame(sys.nframe()))
  return(BrainLists[,2][BrainLists[,1] == gene])
}

coexp.getFriendlyNameForColumnCategories = function(cats){
  friendly.cats = cats
  existing.cats = cats[cats %in% coexp.cell.types]
  
  friendly.cats[cats %in% coexp.cell.types] = 
    coexp.long.friendly.cell.categories[match(existing.cats,coexp.cell.types)]
  return(friendly.cats)
}

#This method is useful for generating cell type enrichment per module from a network
#It will generate a csv ending with celltype.csv with a squared matrix (modules in rows
#and tested cell type markers in columns)
#All can be left to its default value except done for the net.in parameter which
#must have a full path for the network RDS file.
#The csv will be generated at the same folder than the network
#The pdf file with the signals will be generated at plot.file location (can be used as a prefix)
#for the pdf file as in the default
coexp.cellTypeByModule = function(tissue="None",
                                  do.plot=T,
                                  which.one="new",
                                  plot.file=paste0(paste0(gdp.coexp(),"/datasets/nets/cell.type.by.module."),
                                                   tissue,".pdf"),
                                  net.in=NULL,
                                  legend=NULL,
                                  threshold=20,
                                  return.processed=T,
                                  use.grey=F,
                                  display.cats=NULL){ #This last flag FALSE when you want the raw p-values
  
  cat("Entering cellTypeByModue with",which.one,"\n")
  
  if(is.null(net.in)){
    net = coexp.getNetworkFromTissue(tissue=tissue,which.one=which.one)
    modules = coexp.getModulesFromTissue(tissue,which.one)
    is.any.network = F
  }else{
    if(typeof(net.in) == "character")
      net = readRDS(net.in)
    else
      net = net.in
    modules = unique(net$moduleColors)
    is.any.network = T
  }
  if(!use.grey)
    modules = modules[modules != "grey"]
  
  #So the 1st heatmap  
  #will have a column for each cell.types element and a row for
  #each module and we will show -log10(p-values) in a scale
  
  path.ukbec.cell.files = paste0(gdp.coexp(),"/data//")
  if(which.one == "rnaseq"){
    if(tissue == "SNIG")
      enrichment = read.csv(paste0(gdp.coexp(),"data/na/nets/network1237511.4.SNIG.6.rds.USER_terms.csv"),
                            stringsAsFactors=F)
    else
      enrichment = read.csv(paste0(gdp.coexp(),"data/na/nets/network1237918.4.PUTM.6.rds.USER_terms.csv"),
                            stringsAsFactors=F)
  }else if(which.one == "signedrnaseq"){
    if(tissue == "SNIG")
      enrichment = read.csv(paste0(gdp.coexp(),
                                   'data/na/signed/netSNIG.14.4.signed.rds.USER_terms.csv'),
                            stringsAsFactors=F)
    else
      enrichment = read.csv(paste0(gdp.coexp(),"data/na/nets/network1237918.4.PUTM.6.rds.USER_terms.csv"),
                            stringsAsFactors=F)
  }else if(which.one == "unsignedrnaseq"){
    if(tissue == "SNIG")
      enrichment = read.csv(paste0(gdp.coexp(),
                                   'data/na/unsigned/netSNIG.6.4.unsigned.rds.USER_terms.csv'),
                            stringsAsFactors=F)
    else
      enrichment = read.csv(paste0(gdp.coexp(),"data/na/unsigned/netPUTM.6.4.unsigned.rds.USER_terms.csv"),
                            stringsAsFactors=F)
  }else if(which.one == "signedkmeansrnaseq"){
    if(tissue == "SNIG")
      enrichment = read.csv(paste0(gdp.coexp(),
                                   'data/kmwgcna/tests/signed/netSNIG.14.4.signed.rds.kmeans.cor.pca.rds.USER_terms.csv'),
                            stringsAsFactors=F)
    else
      enrichment = read.csv(paste0(gdp.coexp(),
                                   'data/kmwgcna/tests/signed/netPUTM.14.4.signed.rds.kmeans.cor.pca.rds.USER_terms.csv'),
                            stringsAsFactors=F)
  }else if(which.one == "exonic"){
    if(tissue == "SNIG")
      enrichment = read.csv(paste0(coexp.nets[["exonic"]][["SNIG"]],".USER_terms.csv"),
                            stringsAsFactors=F)
    else
      enrichment = read.csv(paste0(coexp.nets[["exonic"]][["PUTM"]],".USER_terms.csv"),
                            stringsAsFactors=F)
  }else if(which.one == "gtexv6"){
    enrichment = read.csv(paste0(coexp.nets[["gtexv6"]][[tissue]],".USER_terms.csv"))
    
  }else if(which.one == "rosmap"){
    e.file = NULL
    if(tissue == "cogdxnotad")
      e.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap/netnotad.8.it.50.rds.USER_terms.csv")
    else if(tissue == "cogdxprobad")
      e.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap/netprobad.11.it.50.rds.USER_terms.csv")
    else if(tissue == "cogdxad")
      e.file = paste0(gdp.coexp(),"/data/rdsnets/rosmap/netad.8.it.50.rds.USER_terms.csv")
    else if(tissue == "all")
      e.file = paste0(gdp.coexp(),"/supplementary/rdsnets/rosmap/netROSMAPSingle.6.it.50.rds.USER_terms.csv")
    if(!is.null(e.file))
      enrichment = read.csv(e.file)
  }else if(which.one == "micro19K"){
    file.name = paste0(paste0(gdp.coexp(),"/supplementary/rdsnets/micro19K/net"),tissue,
                       ".12.signed.it.20.rds.USER_terms.csv")
    #net = readRDS(paste0(paste0(gdp(),"data/kmwgcna/tests/micro19K/net",tissue,
    #		".12.signed.it.20.rds_cor_pca.rds"))
    if(!file.exists(file.name) | is.any.network){
      cat("Generating new user enrichment\n")
      enrichment = userListEnrichment(coexp.fromXtIDToGeneSymbols19K(names(net$moduleColors)),
                                      net$moduleColors, 
                                      fnIn=NULL, useBrainLists = TRUE,
                                      nameOut=file.name, outputCorrectedPvalues=TRUE)
      
    }
    enrichment = read.csv(file.name,stringsAsFactors=F)
  }else if(which.one == "braineac"){
    file.name = paste0(newmic.getNetName(tissue),".USER_terms.csv")
    if(!file.exists(file.name) | is.any.network){
      cat("Generating new user enrichment\n")
      coexp.UserGOenrichment(net.file=getNetworkFromTissue(tissue,"braineac"),
                             net.name=newmic.getNetName(tissue))
    }
    enrichment = read.csv(file.name,stringsAsFactors=F)
  }else if(which.one == "new"){
    file.name = paste0(net.in,".USER_terms.csv")
    #if(!file.exists(file.name)){
    cat("Generating new user enrichment into ",file.name,"\n")
    coexp.UserGOenrichment(net.file=net,
                           net.name=net.in)
    print("Done generating new User enrichment\n")
    #}
    enrichment = read.csv(file.name,stringsAsFactors=F)
  }else{
    stop(paste0("Experiment ",which.one," not known"))
  }
  
  external.ref = read.csv(paste0(path.ukbec.cell.files,"cell_type_TableS1.csv"),stringsAsFactors=F)
  
  all.cell.types = c(coexp.cell.types,coexp.ukbec.cell.types,paste0("External-",colnames(external.ref)))
  cell.type.data = matrix(ncol=length(modules),nrow=length(all.cell.types))
  cell.type.data[,] = 1
  rownames(cell.type.data) = c(coexp.getFriendlyNameForColumnCategories(coexp.cell.types),
                               coexp.ukbec.cell.types,colnames(external.ref))
  colnames(cell.type.data) = modules
  
  
  
  #WGCNA enrichment
  for(module in modules){
    i = match(module,modules)
    for(cell.type in coexp.cell.types){
      j = match(cell.type,coexp.cell.types)
      p.val = as.numeric(enrichment$CorrectedPvalues[enrichment$InputCategories %in% module &
                                                       enrichment$UserDefinedCategories %in% coexp.cell.categories[j]])
      if(length(p.val) > 0)
        cell.type.data[j,i] = p.val #-log10(p.val) 
    }
  }
  
  input.file.names = c(paste0(path.ukbec.cell.files,coexp.ukbec.cell.files),
                       paste0(path.ukbec.cell.files,"cell_type_TableS1.csv.",colnames(external.ref),".txt"))
  cell.types.i.f = c(coexp.ukbec.cell.types,colnames(external.ref))
  
  all.gene.names = coexp.fromAny2GeneName(names(net$moduleColors))
  
  tmp.enr.f = paste0(path.ukbec.cell.files,"cell_type_TableS1.csv.tmp.csv")
  if(file.exists(tmp.enr.f))
    file.remove(tmp.enr.f)
  ukbec.en = userListEnrichment(all.gene.names,net$moduleColors,input.file.names,
                                nameOut=tmp.enr.f)
  if(file.exists(tmp.enr.f)){
    enrichment = read.csv(paste0(path.ukbec.cell.files,"cell_type_TableS1.csv.tmp.csv"),
                          stringsAsFactors=F)
    
    last.cell.types = c(coexp.ukbec.cell.types,colnames(external.ref))
    for(i in 1:nrow(enrichment)){
      module = enrichment$InputCategories[i]
      for(j in 1:length(input.file.names)){
        if(grepl(input.file.names[j],enrichment$UserDefinedCategories[i]))
          break
      }
      category = cell.types.i.f[j]
      cell.type.data[category,module] = enrichment$CorrectedPvalues[i]
    }	
  }
  
  #But before, rename them
  rownames(cell.type.data) = c(coexp.getFriendlyNameForColumnCategories(coexp.cell.types),
                               coexp.ukbec.cell.types,paste0(colnames(external.ref),"-External"))
  unprocessed.cell.type.data = cell.type.data
  cell.type.data = cell.type.data[,apply(cell.type.data,2,function(x){ any(x < 1)}),drop=FALSE]
  cell.type.data = -log10(cell.type.data)
  cell.type.data[is.infinite(cell.type.data)] = max(cell.type.data[!is.infinite(cell.type.data)])
  
  if(threshold > 0)
    cell.type.data[cell.type.data > threshold] = threshold
  #Order by cell type
  
  cell.type.data = cell.type.data[order(rownames(cell.type.data)),,drop=FALSE]
  
  #Now filter 
  if(!is.null(display.cats)){
    cell.type.data = cell.type.data[rownames(cell.type.data) %in% display.cats,]	
  }
  
  
  if(do.plot & (ncol(cell.type.data) >= 2)){
    if(is.null(legend))
      legend = tissue
    if(!is.null(plot.file))
      pdf(plot.file,width=15,height=8)
    heatmap.2(cell.type.data[apply(cell.type.data,1,function(x){ any(x > 2)}),],
              trace="none",
              col=heat.colors(100)[100:1],
              cexCol=0.7,
              cexRow=0.7,
              Rowv=F,
              Colv=T,
              main=paste0("Cell type enrichment for ",legend),
              key=F,srtCol=45,dendrogram="none",
              margins=c(6,20))
    
    if(!is.null(plot.file))
      dev.off()
  }
  if(return.processed)
    return(cell.type.data)
  return(unprocessed.cell.type.data)
}

coexp.genAnnotationCellType = function(tissue="None",
                                       which.one="new",
                                       markerspath,
                                       net.in=NULL,
                                       legend=NULL,
                                       plot.file=NULL,
                                       threshold=20,
                                       return.processed=T,
                                       getMarkerSize=F,
                                       getOnlyOverlap=F,
                                       getMyOwnTest=F,
                                       applyFDR=F,
                                       display.cats=NULL){ #This last flag FALSE when you want the raw p-values
  
  cat("Entering coexp.genAnnotationCellType with",which.one,"\n")
  
  if(which.one == "new"){
    if(typeof(net.in) == "character")
      net = readRDS(net.in)
    else
      net = net.in
  }else{
    net = coexp.getNetworkFromTissue(tissue=tissue,which.one=which.one)
  }
  modules = unique(net$moduleColors)
  
  #So the 1st heatmap  
  #will have a column for each cell.types element and a row for
  #each module and we will show -log10(p-values) in a scale
  
  files = list.files(path=markerspath,full.names = T)
  files = files[grep(pattern=".txt$",files)]
  if(getMarkerSize){
    sizes = lapply(files,function(x){
      nrow(read.delim(x,header=T))
    })
    return(cbind(markerset=files,genes=sizes))
  }
  
  
  markernames = gsub(".txt","",basename(files))
  
  
  ctypes = markernames
  ctypedata = matrix(ncol=length(modules),nrow=length(files))
  ctypedata[,] = 1
  rownames(ctypedata) = ctypes
  colnames(ctypedata) = modules
  
  if(getOnlyOverlap){
    myfunction = Vectorize(function(m,f){
      submarkers = read.delim(f,stringsAsFactors=F,header=T)[,1]
      #cat("Module is",m,"file is",f,"\n")
      #print(names(net$moduleColors)[net$moduleColors == m])
      #print(submarkers)
      return(sum(submarkers %in% names(net$moduleColors)[net$moduleColors == m]))
    })
    
    overlap = outer(modules,files,myfunction)
    colnames(overlap) = gsub(".txt","",basename(files))
    rownames(overlap) = modules
    if(applyFDF){
      overlap = p.adjust(overlap,method="fdr")
    }
    return(overlap)
    
  }
  
  if(getMyOwnTest){
    myfunction = Vectorize(function(m,f){
      submarkers = read.delim(f,stringsAsFactors=F,header=T)[,1]
      ov = sum(submarkers %in% names(net$moduleColors)[net$moduleColors == m])
      nm = sum(net$moduleColors == m)
      
      #cat("Module is",m,"file is",f,"\n")
      #print(names(net$moduleColors)[net$moduleColors == m])
      #print(submarkers)
      return(coexp.testGeneSet(n.module=nm,total.specific=length(submarkers),
                               total.net=length(net$moduleColors),n.module.and.specific=ov)$p.value)
    })
    
    overlap = outer(modules,files,myfunction)
    colnames(overlap) = gsub(".txt","",basename(files))
    rownames(overlap) = modules
    return(overlap)
    
  }
  
  
  
  all.gene.names = names(net$moduleColors)
  all.gene.names = coexp.fromAny2GeneName(names(net$moduleColors))
  
  tmp.enr.f = paste0(markerspath,"enrichment_tmp.csv")
  if(file.exists(tmp.enr.f))
    file.remove(tmp.enr.f)
  ukbec.en = userListEnrichment(all.gene.names,net$moduleColors,files,
                                nameOut=tmp.enr.f)
  if(file.exists(tmp.enr.f)){
    enrichment = read.csv(tmp.enr.f,
                          stringsAsFactors=F)
    
    for(i in 1:nrow(enrichment)){
      module = enrichment$InputCategories[i]
      for(j in 1:length(files)){
        if(grepl(ctypes[j],enrichment$UserDefinedCategories[i]))
          break
      }
      category = ctypes[j]
      ctypedata[category,module] = enrichment$CorrectedPvalues[i]
    }	
  }
  
  uctypedata = ctypedata
  ctypedata = ctypedata[,apply(ctypedata,2,function(x){ any(x < 1)}),drop=FALSE]
  ctypedata = -log10(ctypedata)
  ctypedata[is.infinite(ctypedata)] = max(ctypedata[!is.infinite(ctypedata)])
  
  if(threshold > 0)
    ctypedata[ctypedata > threshold] = threshold
  #Order by cell type
  
  ctypedata = ctypedata[order(rownames(ctypedata)),,drop=FALSE]
  
  
  if((ncol(ctypedata) >= 2)){
    if(is.null(legend))
      legend = tissue
    if(!is.null(plot.file))
      pdf(plot.file,width=15,height=8)
    heatmap.2(ctypedata[apply(ctypedata,1,function(x){ any(x > 2)}),],
              trace="none",
              col=heat.colors(100)[100:1],
              cexCol=0.7,
              cexRow=0.7,
              Rowv=F,
              Colv=T,
              main=paste0("Cell type enrichment for ",legend),
              key=F,srtCol=45,dendrogram="none",
              margins=c(6,20))
    
    if(!is.null(plot.file))
      dev.off()
  }
  if(return.processed)
    return(ctypedata)
  return(uctypedata)
}

coexp.annotateByCellType = function(tissue="None",
                                    do.plot=T,
                                    which.one="new",
                                    plot.file=paste0(paste0(gdp.coexp(),"/datasets/nets/cell.type.by.module."),
                                                     tissue,".pdf"),
                                    net.in=NULL,
                                    legend=NULL,
                                    threshold=20,
                                    return.processed=T,
                                    use.grey=F,
                                    display.cats=NULL){ #This last flag FALSE when you want the raw p-values
  
  cat("Entering coexp.annotateByCellType ",which.one,"\n")
  
  if(is.null(net.in)){
    net = coexp.getNetworkFromTissue(tissue=tissue,which.one=which.one)
    modules = coexp.getModulesFromTissue(tissue,which.one)
    is.any.network = F
  }else{
    if(typeof(net.in) == "character")
      net = readRDS(net.in)
    else
      net = net.in
    modules = unique(net$moduleColors)
    is.any.network = T
  }
  
  if(!use.grey)
    modules = modules[modules != "grey"]
  
  #So the 1st heatmap  
  #will have a column for each cell.types element and a row for
  #each module and we will show -log10(p-values) in a scale
  
  path.ukbec.cell.files = paste0(gdp.coexp(),"/data//")
  if(which.one == "rnaseq"){
    if(tissue == "SNIG")
      enrichment = read.csv(paste0(gdp.coexp(),"data/na/nets/network1237511.4.SNIG.6.rds.USER_terms.csv"),
                            stringsAsFactors=F)
    else
      enrichment = read.csv(paste0(gdp.coexp(),"data/na/nets/network1237918.4.PUTM.6.rds.USER_terms.csv"),
                            stringsAsFactors=F)
  }else if(which.one == "exonic"){
      enrichment = read.csv(paste0(coexp.nets[[which.one]][[tissue]],".USER_terms.csv"),
                            stringsAsFactors=F)
    }else if(which.one == "gtexv6"){
      enrichment = read.csv(paste0(coexp.nets[["gtexv6"]][[tissue]],".USER_terms.csv"))
      
    }else if(which.one == "rosmap"){
      enrichment = read.csv(coexp.ctype$rosmap[[tissue]])
    }else if(which.one == "micro19K"){
      
      enrichment = read.csv(coexp.ctype$micro19K[[tissue]],stringsAsFactors=F)
    }else if(which.one == "new"){
      file.name = paste0(net.in,".USER_terms.csv")
      #if(!file.exists(file.name)){
      cat("Generating new user enrichment into ",file.name,"\n")
      coexp.UserGOenrichment(net.file=net,
                             net.name=net.in)
      print("Done generating new User enrichment\n")
      #}
      enrichment = read.csv(file.name,stringsAsFactors=F)
    }else{
      stop(paste0("Experiment ",which.one," not known"))
    }
    
    external.ref = read.csv(paste0(path.ukbec.cell.files,"cell_type_TableS1.csv"),stringsAsFactors=F)
    
    all.cell.types = c(coexp.cell.types,coexp.ukbec.cell.types,paste0("External-",colnames(external.ref)))
    cell.type.data = matrix(ncol=length(modules),nrow=length(all.cell.types))
    cell.type.data[,] = 1
    rownames(cell.type.data) = c(coexp.getFriendlyNameForColumnCategories(coexp.cell.types),
                                 coexp.ukbec.cell.types,colnames(external.ref))
    colnames(cell.type.data) = modules
    
    
    
    #WGCNA enrichment
    for(module in modules){
      i = match(module,modules)
      for(cell.type in coexp.cell.types){
        j = match(cell.type,coexp.cell.types)
        p.val = as.numeric(enrichment$CorrectedPvalues[enrichment$InputCategories %in% module &
                                                         enrichment$UserDefinedCategories %in% coexp.cell.categories[j]])
        if(length(p.val) > 0)
          cell.type.data[j,i] = p.val #-log10(p.val) 
      }
    }
    
    input.file.names = c(paste0(path.ukbec.cell.files,coexp.ukbec.cell.files),
                         paste0(path.ukbec.cell.files,"cell_type_TableS1.csv.",colnames(external.ref),".txt"))
    cell.types.i.f = c(coexp.ukbec.cell.types,colnames(external.ref))
    
    all.gene.names = coexp.fromAny2GeneName(names(net$moduleColors))
    
    tmp.enr.f = paste0(path.ukbec.cell.files,"cell_type_TableS1.csv.tmp.csv")
    if(file.exists(tmp.enr.f))
      file.remove(tmp.enr.f)
    ukbec.en = userListEnrichment(all.gene.names,net$moduleColors,input.file.names,
                                  nameOut=tmp.enr.f)
    if(file.exists(tmp.enr.f)){
      enrichment = read.csv(paste0(path.ukbec.cell.files,"cell_type_TableS1.csv.tmp.csv"),
                            stringsAsFactors=F)
      
      last.cell.types = c(coexp.ukbec.cell.types,colnames(external.ref))
      for(i in 1:nrow(enrichment)){
        module = enrichment$InputCategories[i]
        for(j in 1:length(input.file.names)){
          if(grepl(input.file.names[j],enrichment$UserDefinedCategories[i]))
            break
        }
        category = cell.types.i.f[j]
        cell.type.data[category,module] = enrichment$CorrectedPvalues[i]
      }	
    }
    
    #But before, rename them
    rownames(cell.type.data) = c(coexp.getFriendlyNameForColumnCategories(coexp.cell.types),
                                 coexp.ukbec.cell.types,paste0(colnames(external.ref),"-External"))
    unprocessed.cell.type.data = cell.type.data
    cell.type.data = cell.type.data[,apply(cell.type.data,2,function(x){ any(x < 1)}),drop=FALSE]
    cell.type.data = -log10(cell.type.data)
    cell.type.data[is.infinite(cell.type.data)] = max(cell.type.data[!is.infinite(cell.type.data)])
    
    if(threshold > 0)
      cell.type.data[cell.type.data > threshold] = threshold
    #Order by cell type
    
    cell.type.data = cell.type.data[order(rownames(cell.type.data)),,drop=FALSE]
    
    #Now filter 
    if(!is.null(display.cats)){
      cell.type.data = cell.type.data[rownames(cell.type.data) %in% display.cats,]	
    }
    
    
    if(do.plot & (ncol(cell.type.data) >= 2)){
      if(is.null(legend))
        legend = tissue
      if(!is.null(plot.file))
        pdf(plot.file,width=15,height=8)
      heatmap.2(cell.type.data[apply(cell.type.data,1,function(x){ any(x > 2)}),],
                trace="none",
                col=heat.colors(100)[100:1],
                cexCol=0.7,
                cexRow=0.7,
                Rowv=F,
                Colv=T,
                main=paste0("Cell type enrichment for ",legend),
                key=F,srtCol=45,dendrogram="none",
                margins=c(6,20))
      
      if(!is.null(plot.file))
        dev.off()
    }
    if(return.processed)
      return(cell.type.data)
    return(unprocessed.cell.type.data)
  }